File: Week04/Mod01-dsuminmax.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll  long long int
#define vi  vector<int>
#define vll vector<long long int>

class UnionFind {
    private:
        vi p, rank, setSize, maxSet, minSet;
        int numSets;
    public:

        // Initialization

        UnionFind(int N){
            // The goal here is to create N singleton sets.

            p.assign(N, 0);
            // To begin with, everyone is their own parent:
            for(int i = 0; i < N; i++)
                p[i] = i;
            
            // the height of all trees are zero:
            rank.assign(N, 0);

            // the sizes of the individual sets are one:
            setSize.assign(N, 1);

            // the maximum and minimum element in the set is the element itself:

            maxSet.assign(N, 0);
            minSet.assign(N, 0);
            for(int i = 0; i < N; i++){
                maxSet[i] = i;            
                minSet[i] = i;            
            }

            // the number of sets is N:
            numSets = N;
        }

        // FindSet(i): return the representative element of the set that i belongs to.

        int findSet(int i){
            // if already at a representative element,
            // signified by the fact that the parent pointer points to the element itself,
            // return;

            if(p[i] == i)
                return i;

            // otherwise find the representative of the parent,
            // but also perform path compression by pointing the current element
            // to the final representative:

            else
                return p[i] = findSet(p[i]);
        }

        // isSameSet(i,j): returns true if and only if i and j belong to the same set.

        bool isSameSet(int i, int j){
            // Observe that i and j belong to the same set
            // if and only if they have the same representative.
            return findSet(i) == findSet(j); 
        }

        // sizeOfSet(i): returns the size of the set that the element i belongs to.

        int sizeOfSet(int i){
            return setSize[findSet(i)];
        }
        
        // numDisjointSets(): returns the number of sets being maintained currently.

        int numDisjointSets(){
            return numSets; 
        }
        
        // maxInSet(): returns the maximum element of the set

        int maxInSet(int i){
            return maxSet[findSet(i)]; 
        }
        
        // minInSet(): returns the minimum element of the set

        int minInSet(int i){
            return minSet[findSet(i)]; 
        }

        void unionSet(int i, int j){
            if(isSameSet(i,j)) 
                return; 
            
            int x = findSet(i);
            int y = findSet(j);

            if(rank[x] > rank[y])
                swap(x,y);
            
            p[x] = y; 

            if (rank[x] == rank[y])
                rank[y]++;
            
            setSize[y] += setSize[x];
            maxSet[y] = max(maxSet[x],maxSet[y]);
            minSet[y] = min(minSet[x],minSet[y]);
            numSets--;

        }


};


int main(){
    #ifndef ONLINE_JUDGE 
        freopen("in.txt", "r", stdin);
        freopen("out.txt", "w", stdout);
    #endif

    std::ios::sync_with_stdio(false);

    int n, m;
    cin >> n >> m;
    
    UnionFind UF(n+1);

    while(m--){
        string S;
        cin >> S;


        if(S == "union"){
            int x,y;
            cin >> x >> y;
            UF.unionSet(x,y);
        }          
        else{
            int x;
            cin >> x;
            cout << UF.minInSet(x) << " " << UF.maxInSet(x) <<  " " << UF.sizeOfSet(x) << endl;
        }
    }

    return 0;
}

File: Week04/Mod01-dsuminmax.py
# Problem Link: https://codeforces.com/edu/course/2/lesson/7/1/practice/contest/289390/problem/B

import sys
input = sys.stdin.readline

class DSU:
  def __init__(self, n):
    """Initializes the data structure with 'n' singleton sets"""
    
    # Every element is its own parent
    self.parent = list(range(n))

    # The height of all trees is 0 (an upper bound of depth)
    self.rank = [0] * n

    # The size of each set is 1 (singleton sets)
    self.setSize = [1] * n

    # Each set's minimum and the maximum element is the one element in it
    self.minSet = list(range(n))
    self.maxSet = list(range(n))

    # There are n singleton sets initially
    self.setCount = n

  def findSet(self, i):
    """Returns the parent element of the set that 'i' belongs to"""
    
    # If the element is already a parent element of a set, return it
    if self.parent[i] == i:
      return i
    
    # Else, move up the tree to find the parent, and perform path compression
    self.parent[i] = self.findSet(self.parent[i])

    # Return the parent
    return self.parent[i]

  def sameSet(self, i, j):
    """Checks if the two elements belong to the same set"""
    return self.findSet(i) == self.findSet(j)

  def union(self, i, j):
    """Combines the sets that 'i' and 'j' belong to"""

    # if the elements belong to the same set, we don't need to do anything
    # If they're not, we perform the union operation
    if not self.sameSet(i, j):

      # Find the parent elements of the two elements
      x = self.findSet(i)
      y = self.findSet(j)

      # Make sure that 'x' is the tree with the smaller rank (proxy to depth)
      x, y = (y, x) if self.rank[x] > self.rank[y] else (x, y)

      # Set y as the parent of x
      self.parent[x] = y

      # If the ranks were the same, then the depth of the final tree must increase by 1
      if self.rank[x] == self.rank[y]:
        self.rank[y] += 1
      
      # Update the size of the combined set
      self.setSize[y] += self.setSize[x]

      # Update the min and max values of the combined set
      self.minSet[y] = min(self.minSet[y], self.minSet[x])
      self.maxSet[y] = max(self.maxSet[y], self.maxSet[x])

      # Lastly, reduce the set count, since we merged two sets into one
      self.setCount -= 1

  def sizeOfSet(self, i):
    """Returns the size of the set that the given element belongs to"""
    return self.setSize[self.findSet(i)]
  
  def getSetCount(self):
    """Returns the current count of disjoint sets"""
    return self.setCount

  def minInSet(self, i):
    """Returns the minimum value of the set that 'i' belongs to"""
    return self.minSet[self.findSet(i)]

  def maxInSet(self, i):
    """Returns the maximum value of the set that 'i' belongs to"""
    return self.maxSet[self.findSet(i)]

if __name__ == '__main__':
  n, m = (int(i) for i in input().split())

  disjointSet = DSU(n+1)

  for _ in range(m):
    query = input().split()

    if len(query) == 3:
      u, v = int(query[1]), int(query[2])
      disjointSet.union(u, v)
    else:
      v = int(query[1])
      print(disjointSet.minInSet(v), disjointSet.maxInSet(v), disjointSet.sizeOfSet(v))